src/main/java/org/opentripplanner/routing/algorithm/raptor/transit/mappers/TransitLayerMapper.java

```diff
@@ -66,18 +81,127 @@ private TransitLayer map() {
+        }
+
+        // Convert all old TripPatterns into new ones, keeping a Map between the two.
+        // Do this conversion up front (rather than lazily on demand) to ensure pattern IDs match the sequence of
+        // patterns in source data.
+        Map<org.opentripplanner.routing.edgetype.TripPattern, TripPattern> newTripPatternForOld = new HashMap<>();
+        int patternId = 0;
+        for (org.opentripplanner.routing.edgetype.TripPattern oldTripPattern : graph.tripPatternForId.values()) {
+            TripPattern newTripPattern = new TripPattern(
+                    patternId++,
+                    // TripPatternForDate should never access the tripTimes inside the TripPattern, so I've left them null.
+                    // No TripSchedules in the pattern itself; put them in the TripPatternForDate
+                    null,
+                    oldTripPattern.mode,
+                    listStopIndexesForTripPattern(oldTripPattern, indexByStop)
+            );
+            newTripPatternForOld.put(oldTripPattern, newTripPattern);
+        }
+
+        // Presumably even when applying realtime, many TripTimes will recur on future dates.
+        // This Map is used to deduplicate the resulting TripSchedules.
+        Map<TripTimes, TripSchedule> tripScheduleForTripTimes = new HashMap<>();
+
+        // The return value of this entire process.
+        HashMap<LocalDate, List<TripPatternForDate>> tripPatternsForDates = new HashMap<>();
+
+        // Loop over all dates for which any service is defined
+        // We need the original TripPattern to look it up in the timetable snapshot, but also need to convert it to a
+        // new Raptor TripPattern.
+        // Get a frozen snapshot of timetable data so it isn't changed by incoming realtime data while we're iterating.
+        TimetableSnapshot timetableSnapshot = null;
+        if (graph.timetableSnapshotSource == null) {
+            LOG.info("There is no timetable snapshot source. This TransitLayer will reflect scheduled service.");
+        } else {
+            timetableSnapshot = graph.timetableSnapshotSource.getTimetableSnapshot();
+            LOG.info("Got timetable snapshot. This TransitLayer will reflect realtime updates.");
+        }
+        Set<ServiceDate> allServiceDates = serviceIdsForServiceDate.keySet();
+        Map<Timetable, List<TripTimes>> sortedTripTimesForTimetable = new HashMap<>();
+        for (ServiceDate serviceDate : allServiceDates) {
+            // Create LocalDate equivalent to the OTP/GTFS ServiceDate object, serving as the key of the return Map.
+            LocalDate localDate = ServiceCalendarMapper.localDateFromServiceDate(serviceDate);
+            // Create a List to hold the values for one entry in the return Map.
+            List<TripPatternForDate> values = new ArrayList<>();
+            // This nested loop could be quite inefficient.
+            // Maybe determine in advance which patterns are running on each service and day.
```

Comment:
Is this comment ment for the next 2 loops, the first loop? or the 2nd loop? The phrase "patterns" indicate it apply to the 2nd loop.

Reply:
Yes, it applies to the second loop. I must have inserted the serviceId loop after writing the comment. I can move it. I do think this could be a source of inefficiency but would need to profile a bit.

Reply:
I fixed it.